### ISSUE TITLE
[BUG] Kibana - Findings - "CIS Section" column sorting is case sensitive

### ISSUE THREAD

**Issue by uri-weisman** [elastic/cloudbeat#104]
**Describe the bug**
When clicking the `CIS Section` column for sorting, results that start in lower case appears on the last page. 

**Preconditions**
Kibana + ES + CIS k8s benchmark integration is installed.
Elastic Agent deployed on a K8s cluster.

**To Reproduce**

1. Open Kibana and navigate to the Cloud Security section.
2. Go to the Findings page.
3. Click on the `CIS Section` for sorting.
4. Navigate to the last page of results.

**Expected behavior**
Results should be sorted alphabetically regardless of their first letter case.

**Screenshots**
![Screen Shot 2022-05-10 at 15 07 34](https://user-images.githubusercontent.com/68195305/167624819-346722bf-005e-45bd-93da-1e4e4933a18b.png)

DOD -
- [x] Fixing the bug
- [x] functional test

KIBANA VERSION: 8.3.0
BUILD: 52512
COMMIT: df225b213b188c81888141cee2ec191424fc0649

**Comment by kfirpeled** [comment]
This is true for `Resource Type` and `Resource Name` columns as well

**Comment by kfirpeled** [comment]
This might require changing the mappings and adding `analyzer`/ `normalizer`

**Comment by ofiriro3** [comment]
As far as I understand we cannot change the sorting function easily and it would require some investigation and changes in our code.

![image](https://user-images.githubusercontent.com/19221736/196887134-1aaa8b11-695c-45f2-82cc-48129b22065b.png)

Moreover, all the EUI tables work that way, sorting first by a capital letter and then by a lowercase letter - are we really certain that the following is a bug? would it make sense to change the behavior of certain columns only?

@kfirpeled @tehilashn @tinnytintin10 @eyalkraft @JordanSh

**Comment by kfirpeled** [comment]
From https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-normalizers.html

```
Elasticsearch ships with a `lowercase` built-in normalizer. For other forms of normalization a custom configuration is required.
```

**Comment by ofiriro3** [comment]
As I investigate I'll document possible solutions that I've found:

**Using normalizers**
The normalizer creates a single token, which is the normalized value of the indexed field.
The normalizer is applied prior to indexing the keyword, as well as at search-time when the keyword field is searched via a query parser such as the [match](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query.html) query or via a [term](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-term-query.html)-level query such as the term query.

Diving a bit deep into the technical aspect of it, when indexing a value using a normalizer the "Raw" document will remain the same, but the indexed field will be the normalized one.

For example:
```
{
  "_index": "ofir-test-5",
  "_id": "1vgdBIQBv5i7gqlnNO-V",
  "_version": 1,
  "_score": 0,
  "_source": {
    "foo": "AAAA"
  },
  "fields": {
    "foo": [
      "aaaa"
    ]
  }
}
```

**Comment by ofiriro3** [comment]
**Using a custom analyzer**
The analyzer parameter specifies the [analyzer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analyzer-anatomy.html) used for [text analysis](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis.html) when indexing or searching a text field.
An important note here is that analyzers only work on `text` type fields and not on keywords.
In order to enable this option, one must enable `fieldData` in its mapping as `test` fields are not available for sorting/scripting and aggregation. The field data solution is not preferable one as it may consume a lot of heap space. 

The advantage of this solution is that it keeps the indexed field value as it is and does not lowercasing it.

```json
{
  "_index": "ofir-test-analyzer-fielddata",
  "_id": "i_lKBIQBv5i7gqlnDxNy",
  "_version": 1,
  "_score": 0,
  "_source": {
    "foo": "AAAAA"
  },
  "fields": {
    "foo": [
      "AAAAA"
    ]
  }
}
```

In order to implement this solution we will need to change the mapping and add the analyzers to the relevant fields.
For example, something like the following:
```
PUT ofir-test-analyzer-fielddata
{
  "settings": {
     "analysis": {
       "analyzer": {
         "custom_sort_analyzer": {
            "filter": ["lowercase"],
            "tokenizer": "keyword"
            
        }
       }
     }
  },
     "mappings": {
    "properties": {
      "foo": {
        "type": "text",
        "analyzer": "custom_sort_analyzer",
        "fielddata": true
      }
    }
  }
}
```

**Comment by ofiriro3** [comment]
**Duplicating the field**
Another option that we can consider is to duplicate this field and then to use the sort on normalized field, and show the result of the original field.

Something like the following will work:
```
PUT /test_index/
{
  "settings": {
    "analysis": {
      "normalizer": {
        "myLowercase": {
          "type": "custom",
          "filter": [ "lowercase" ]
        }
      }
    }
  },
  "mappings":{
     "post":{
        "properties":{
           "foo":{
              "normalizer":"myLowercase",
              "type":"keyword",
              "copy_to": [
              "foo2"
            ]
           },
           "foo2":{
              "type":"keyword"
           }
        }
     }
  }
}
```

**Comment by kfirpeled** [comment]
Summerizing what we've discussed in the morning sync, let's try the latter option with multi-field mapping. And check if it works when upgrading from an existing integration.

**Comment by ofiriro3** [comment]
After looking at the Kibana code base I see that the normalizer is well-used across all the plugins.

I also posted a question to the `Kibana` team via slack, to see what they think and what are the best practices in these cases.

As far as I see both the normalizer and the duplicate fields will be easy to implement, I tend to go with the normalizer since it looks cleaner to me and requires fewer code changes.
One of the most popular normalizers is the `lowercase` normalizer which was introduced exactly for cases similar to ours.

>The normalizer is applied prior to indexing the keyword, as well as at search time when the keyword field is searched via a query parser such as the [match](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query.html) query or via a [term](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-term-query.html)-level query such as the term query.


If we will ever remove the `_source` parameter, I think that moving from a `normalizer` solution to the `duplicate field` would not be complex. 

That being said, if we are certain that the `_source` field will be removed very soon, so maybe it is the best option.

**Comment by ofiriro3** [comment]
@kfirpeled I have just checked it and the index is being updated right after the package upgrade, no need to wait for a new index creation.

**Comment by eyalkraft** [comment]
A possible approach that could simplify things would be to have our integration auto updated
- https://github.com/elastic/kibana/issues/134247

**Comment by ofiriro3** [comment]
We decided to postpone the implementation of this task since we would like to evaluate all options and their implications on our system.

I would like to summarize the correct status before moving on to other tasks.
We decided to give up on `Using a custom analyzer` since it involves changing the field type from `keyword` to `text`. This change may require some complex backward/forward compatibility support that will not be easy to provide.

Currently, there are three approaches that we consider:

1. [Using normalizers](https://github.com/elastic/cloudbeat/issues/104#issuecomment-1288061570) - It will require changing the current mapping of `rule.section`, and adding a normalizer to the field mapping.
**ADV:**
a. Won't require any Kibana changes
b. Easy to implement.
**DSV**
a. If we will ever decide to remove the `_source` we will need to copy the original field to another field/remove the normalizer.
b. Dangling state - It will only work when a new index will be created.

2. [Duplicating the field](https://github.com/elastic/cloudbeat/issues/104#issuecomment-1288078947) - We can use the ingest pipeline in order to duplicate the field and create a lowercased version of it. 
**ADV:**
a. We only extend the current mapping (no removals).
b. We can easily remove the `_source` field.
**DSV**
a. A bit complex when it comes to backward/forward compatibility, Kibana will have to provide a call-back method in case the new field does not exist.

4. Ingest pipeline - This is a new solution that we only came up with yesterday. We can create an ingest pipeline that will camel-case all the required fields.
**ADV**
a. Easy to implement
b. We can easily remove the `_source` field.
**DSV**
a. We are changing the original data.

5. Cloudbeat/Rules change - As @eyalkraft mentioned we could just edit the rule text such that the data will always start with an upper-case letter.


@eyalkraft @kfirpeled @oren-zohar

**Comment by kfirpeled** [comment]
Thanks @ofiriro3 for the summary,

As for option 3, worth adding an advantage it also gives backward compatibility for older versions of cloudbeat, although, not older versions of the the installed package

A 5th option to bring up to the table is a runtime field using `painless` script to sort by. You can see an example [here](https://stackoverflow.com/a/64328978)
The advantage here, it will work for all the use cases but might take a hit in performance (which I truly believe we can live with, since we don't have a large index we search on, under 50MB)

**Comment by kfirpeled** [comment]
Unblocking issue, https://github.com/elastic/kibana/issues/144525 has been merged so it is safe to push this for 8.6

**Comment by ofiriro3** [comment]
I am adding to the DOD of the ticket the addition of functional tests.

fyi @kfirpeled

**Comment by ofiriro3** [comment]
I am putting this item on `block`.

We would like to add a functional test before merging it.
Once this one will be done, I'll work on the FTR test for it https://github.com/elastic/kibana/issues/147998
